sah.8
.TH SAH 8 "February 12, 2019" "version 0.9.4" "SAH Manual"
.SH NAME
Simple AUR Helper (SAH) and Pacman wrapper
.SH SYNOPSIS
.B sah
<operation> [packages]
.SH DESCRIPTION
Simple AUR Helper and Pacman wrapper
.SH OPERATIONS
.B -S, install
.RS
Install package/packages from AUR
.RE
.B -Sp
.RS
Install package/packages from Pacman
.RE
.B -Sf
.RS
Install packages from file with list of packages (AUR)
.RE
.B -Spf
.RS
Install packages from file with list of packages (Pacman)
.RE
.B -U
.RS
Install local or remote package
.RE
.B -Syy, update
.RS
Update package database
.RE
.B -Syu, upgrade
.RS
Update installed packages (Pacman + AUR)
.RE
.B -Sc, autoremove
.RS
Clean the package cache
.RE
.B -R, remove
.RS
Remove package/packages
.RE
.B -Rs, purge
.RS
Remove package/packages with dependencies which aren't required by any other installed packages
.RE
.B -Ro
.RS
Remove packages-orphans
.RE
.B -Q, list
.RS
Show installed packages (All)
.RE
.B -Qe
.RS
Show installed packages (Explicitly)
.RE
.B -Qm
.RS
Show installed packages (AUR)
.RE
.B -Ss
.RS
Search for packages in the database
.RE
.B -Qs
.RS
Search for already installed packages
.RE
.B -Si
.RS
Show information about package
.RE
.B -Qi
.RS
Show information about package (for locally installed packages)
.RE
.B -Qdt
.RS
Show all packages no longer required as dependencies (orphans)
.RE
.B -V, --version
.RS
Display version and exit
.RE
.B -h, --help
.RS
Display SAH built-in manual
.RE
.SH OPTIONS
.B custom [operation]
.RS
Execute custom Pacman operation
.RE
.B stat
.RS
Show packages statistics
.RE
.B config
.RS
Open SAH config file via selected text editor
.RE
.B resetconfig
.RS
Reset to default SAH config file from GitHub repo
.RE
.B pacconf
.RS
Open Pacman config file via selected text editor
.RE
.B mirrorlist
.RS
Open mirrorlist file via selected text editor
.RE
.B updatemirrors
.RS
Update mirrorlist from Arch Linux website
.RE
.B changelog
.RS
Show SAH changelog
.RE
.B top
.RS
Show AUR TOP-10 packages
.RE
.B top [1-10]
.RS
Install package from AUR TOP-10
.RE
.B cache [package1] [package2] ...
.RS
Cache AUR package/packages
.RE
.B cache install [package1] [package2] ...
.RS
Install AUR package/packages from cache
.RE
.B cache update [package1] [package2] ...
.RS
Update cached AUR package/packages
.RE
.B cache update --all
.RS
Update all cached AUR packages
.RE
.B cache remove [package1] [package2] ...
.RS
Remove AUR package/packages from cache
.RE
.B cache list
.RS
Show all cached AUR packages
.RE
.B cache clean
.RS
Remove all cached AUR packages
.RE
.B browse
.RS
Browse all packages from Arch Linux website via web browser
.RE
.B browse [package]
.RS
Browse the package from Arch Linux website via web browser
.RE
.B view [package]
.RS
View PKGBUILD of AUR package
.RE
.SH CONFIGURATION
You can edit SAH config file to set up some settings
.PP
SAH config file path:
.B /etc/sah_config
.PP
Also you can use 'sah config' to open SAH config file via selected text editor
.PP
.B Supported properties in config:
.PP
.B editor (package)
.RS
text editor for editing configs
.RE
.B viewer (package)
.RS
text viewer for viewing text files
.RE
.B browser (package)
.RS
web browser to browse packages from Arch Linux website
.RE
.B update_pacman (true/false)
.RS
enable/disable updating of Pacman packages
.RE
.B update_aur (true/false)
.RS
enable/disable updating of AUR packages
.RE
.B aur_update_notify (true/false)
.RS
notify about new versions of AUR packages during updating
.RE
.B auto_cache (true/false)
.RS
enable/disable auto caching for AUR packages
.RE
.B aur_update_ignore (package1,package2,...)
.RS
skip updating of some AUR packages
.RE
.B mirrorlist_country (country code)
.RS
mirrors country
.RE
.B mirrorlist_protocol (http/https)
.RS
mirrors protocol
.RE
.B mirrorlist_ip_version (4/6)
.RS
mirrors IP version
.RE
.B rmd (true/false)
.RS
remove make dependencies of AUR packages during installation or updating
.RE
.B pgp_check (true/false)
.RS
enable/disable verifying PGP signatures of source files
.RE
.B needed (true/false)
.RS
enable/disable reinstalling packages if they are already up-to-date
.RE
.B noconfirm (true/false)
.RS
enable/disable waiting for user input before proceeding with operations
.RE
.PP
.B Properties examples:
.PP
editor=nano
.PP
viewer=less
.PP
browser=firefox
.PP
update_pacman=true
.PP
update_aur=true
.PP
aur_update_notify=false
.PP
auto_cache=true
.PP
aur_update_ignore=yay,dropbox,google-chrome
.PP
mirrorlist_country=RU
.PP
mirrorlist_protocol=http
.PP
mirrorlist_ip_version=4
.PP
rmd=false
.PP
pgp_check=false
.PP
needed=false
.PP
noconfirm=false
.SH EXAMPLES
Install package/packages from AUR
.RS
.B sah
-S [package1] [package2] ...
.RE
Install package/packages from Pacman
.RS
.B sah
-Sp [package1] [package2] ...
.RE
Install local package
.RS
.B sah
-U /path/to/package/package_name-version.pkg.tar.xz
.RE
Install remote package
.RS
.B sah
-U https://www.example.com/repo/example.pkg.tar.xz
.RE
Update installed packages (Pacman + AUR)
.RS
.B sah
-Syu
.RE
Execute custom Pacman operation (-Syyu)
.RS
.B sah
custom -Syyu
.RE
Execute custom Pacman operation (-Syu and install a package)
.RS
.B sah
custom -Syu chromium
.RE
Install package from AUR TOP-10 (7th)
.RS
.B sah
top 7
.RE
Cache package from AUR
.RS
.B sah
cache spotify
.RE
Update cached packages from AUR
.RS
.B sah
cache update google-chrome dropbox
.RE
Browse all AUR packages via web browser
.RS
.B sah
browse --aur
.RE
Search for packages (Pacman) via web browser
.RS
.B sah
browse --pacman linux
.RE
View PKGBUILD of AUR package
.RS
.B sah
view yay
.RE
.SH AUTHOR
Stepan Skryabin
.SH "SEE ALSO"
pacman(8)


changelog.txt
v0.9.4 (21.03.2021)
- Added a feature to remove packages-orphans

v0.9.3 (18.01.2021)
- Updated links to Arch Linux website

v0.9.2 (05.09.2020)
- Changed way of getting AUR TOP-10 packages

v0.9.1 (01.02.2020)
- Added showing time of AUR TOP-10 packages update

v0.9 (01.02.2020)
- Added cache for AUR packages
- Added optional auto caching for AUR packages
- Added a feature to view PKGBUILD of AUR packages
- Added config option to select text viewer
- Removed logging
- Updated AUR TOP-10 packages

v0.8.2 (10.12.2019)
- Added a feature to browse packages from Arch Linux website via web browser
- Updated AUR TOP-10 packages

v0.8.1 (21.07.2019)
- Updated SAH documentation

v0.8 (20.07.2019)
- Added config option to select editor
- Added a feature to show packages statistics
- Added a feature to show all installed packages
  -Qe is for explicitly installed packages from now.
  Also added APT-style operation.
- Disabled logging by default
- Execute all Pacman operations via sudo
- Paths and variables improvements
- Updated AUR TOP-10 packages

v0.7.3 (04.07.2019)
- Moved AUR TOP-10 packages to separate file
- Updated AUR TOP-10 packages

v0.7.2 (20.06.2019)
- Added a feature to install packages from file with list of packages
- Added a feature to enable/disable waiting for user input before proceeding with operations
- Changed SAH option 'updateconfig' to 'resetconfig'
- Updated AUR TOP-10 packages

v0.7.1 (16.06.2019)
- Updated paths to some files
- Updated AUR TOP-10 packages

v0.7 (14.06.2019)
- New features:
  - Update package database
  - Install local or remote package
  - Execute custom Pacman operation
- Added some APT-style operations
- Added AUR TOP-10 popular packages
- Updated SAH website

v0.6 (02.05.2019)
- New features:
  - Notify about new versions of AUR packages during updating
  - Disable updating of Pacman packages
  - Disable updating of AUR packages
  - Open Pacman config file via nano editor
  - Open mirrorlist file via nano editor
  - Update mirrorlist from Arch Linux website
  - Clear SAH log file
  - Update SAH config file from GitHub repo
- Show count of query results (-Qe/-Qm/-Qdt)
- Added index.html for GitHub Pages

v0.5.4 (27.04.2019)
- Fixed SAH man page

v0.5.3 (27.04.2019)
- Probably, I found the right solution to fix PKGBUILD

v0.5.2 (27.04.2019)
- Another try to fix config installation in PKGBUILD

v0.5.1 (27.04.2019)
- Tried to fix config installation in PKGBUILD

v0.5 (27.04.2019)
- New features:
  - Open SAH config file via nano editor
  - Show SAH changelog
  - Enable/disable reinstalling packages if they are already up-to-date
- Added 'debug' option for test purposes
- Added logging for operations and options
- Added Bash completion for SAH options
- Fixed default config installation during SAH updating

v0.4.3 (26.04.2019)
- Added a feature to enable/disable verifying PGP signatures of source files
- Some code improvements and fixes

v0.4.2 (21.03.2019)
- Fixed SAH man page

v0.4.1 (21.03.2019)
- Added config file for SAH
- Removed --rmd as an option of SAH command.
Since v0.4.1 you need set up 'true' rmd property in config file to remove make dependencies.
- Added a feature to skip updating of some AUR packages

v0.4 (15.03.2019)
- Added man page for SAH

v0.3.2 (15.03.2019)
- Fixed cases when --rmd is not the last option

v0.3.1 (19.02.2019)
- Fixed update checking for versions with single quotes
- Added showing numbers of checking AUR packages

v0.3 (18.02.2019)
- New features:
  - Remove packages with their dependencies
  - Search for packages
  - Show information about packages
  - Show all packages no longer required as dependencies
- SAH info shows via less

v0.2 (16.02.2019)
- Changed paths to files containing lists of installed AUR packages
- Added a feature to clean the package cache
- Code optimization

v0.1.2 (15.02.2019)
- Fixed updating SAH when installed latest SAH version

v0.1.1 (15.02.2019)
- Added updating SAH when packages update

v0.1 (15.02.2019)
- Release on GitHub


sah.sh
#!/bin/bash

### Simple AUR Helper (SAH) and Pacman wrapper
VERSION="0.9.4"

##### Settings

# Variables to paths
pkg_list_path="$HOME/.sah/sah_pkg_list"
pkg_list_path_v="$HOME/.sah/sah_pkg_list_v"
PKGBUILDs_path="/tmp/PKGBUILDs"
SAH_cache_path="$HOME/.sah/cache"
SAH_cache_list_path="$HOME/.sah/cache_list"
SAH_config_path="/etc/sah_config"
pacman_config_path="/etc/pacman.conf"
mirrorlist_path="/etc/pacman.d/mirrorlist"
SAH_help_path="/usr/share/sah/help"
SAH_changelog_path="/usr/share/sah/changelog"
aur_top_path="/tmp/aur_top"

kernel_version=$(uname -r)
date_time_format=$(date +"%d.%m.%Y %H:%M:%S")

# Reading config options
read_config() {
  cat $SAH_config_path | grep $1 | awk -F "=" '{print $2}'
}

SAH_editor=$(read_config editor)
SAH_viewer=$(read_config viewer)
SAH_browser=$(read_config browser)

update_pacman_check=$(read_config update_pacman)
update_aur_check=$(read_config update_aur)
aur_update_notify_check=$(read_config aur_update_notify)
auto_cache_check=$(read_config auto_cache)

mirrorlist_country=$(read_config mirrorlist_country)
mirrorlist_protocol=$(read_config mirrorlist_protocol)
mirrorlist_ip_version=$(read_config mirrorlist_ip_version)

rmd_check=$(read_config rmd)
pgp_check=$(read_config pgp_check)
needed_check=$(read_config needed)
noconfirm_check=$(read_config noconfirm)

# Remove make dependencies (-si/-sir)
if [[ $rmd_check == "false" ]]; then
  makepkg_type="-si"
elif [[ $rmd_check == "true" ]]; then
  makepkg_type="-sir"
fi

# PGP check (--skippgpcheck)
if [[ $pgp_check == "false" ]]; then
  makepkg_type="$makepkg_type --skippgpcheck"
elif [[ $pgp_check == "true" ]]; then
  makepkg_type="$makepkg_type"
fi

# Needed (--needed)
# Don't reinstall packages if they are already up-to-date
if [[ $needed_check == "false" ]]; then
  makepkg_type="$makepkg_type"
elif [[ $needed_check == "true" ]]; then
  makepkg_type="$makepkg_type --needed"
fi

# No confirm (--noconfirm)
# Don't wait for user input before proceeding with operations
if [[ $noconfirm_check == "false" ]]; then
  makepkg_type="$makepkg_type"
elif [[ $noconfirm_check == "true" ]]; then
  makepkg_type="$makepkg_type --noconfirm"
fi

##### Functions

##### Main code

# SAH Install
if [[ $1 == "-S" || $1 == "install" ]]; then
  aur_pkg_range="${@:2}"
  for aur_pkg in $aur_pkg_range
  do
    git clone https://aur.archlinux.org/$aur_pkg.git
    if [[ $auto_cache_check == "true" ]]; then
      test -d $SAH_cache_path/$aur_pkg
      if [[ $? == "0" ]]; then
        cd $SAH_cache_path/$aur_pkg
        git pull
        cd - > /dev/null
      elif [[ $? == "1" ]]; then
        git clone -q https://aur.archlinux.org/$aur_pkg.git $SAH_cache_path/$aur_pkg
      fi
    fi
    cd $aur_pkg
    echo "Installing $aur_pkg..."
    makepkg $makepkg_type
    cd ..
    rm -rf $aur_pkg
  done
# SAH Install Pacman
elif [[ $1 == "-Sp" ]]; then
  sudo pacman -S ${@:2}
# SAH Install Package List
elif [[ $1 == "-Sf" || $1 == "-Spf" ]]; then
  if [[ $2 == "" ]]; then
    if [[ $1 == "-Sf" ]]; then
      install_package_list_message="Enter the path to the file with list of packages for installation (AUR): "
    elif [[ $1 == "-Spf" ]]; then
      install_package_list_message="Enter the path to the file with list of packages for installation (Pacman): "
    fi
    read -e -p "$install_package_list_message" install_pkg_list
  elif [[ $2 != "" ]]; then
    install_pkg_list=$2
  fi

  list_items_count=$(cat $install_pkg_list | wc -l)
  list_line=1
  pkg_string=""
  for (( i = 0; i < $list_items_count; i++ )); do
    list_item=$(sed -n ${list_line}p $install_pkg_list)
    if [[ $i == 0 ]]; then
      pkg_string="$list_item"
    else
      pkg_string="$pkg_string $list_item"
    fi
    list_line=$(($list_line + 1))
  done

  if [[ $1 == "-Sf" ]]; then
    aur_pkg_range=$pkg_string
    for aur_pkg in $aur_pkg_range
    do
      git clone https://aur.archlinux.org/$aur_pkg.git
      if [[ $auto_cache_check == "true" ]]; then
        test -d $SAH_cache_path/$aur_pkg
        if [[ $? == "0" ]]; then
          cd $SAH_cache_path/$aur_pkg
          git pull
          cd - > /dev/null
        elif [[ $? == "1" ]]; then
          git clone -q https://aur.archlinux.org/$aur_pkg.git $SAH_cache_path/$aur_pkg
        fi
      fi
      cd $aur_pkg
      echo "Installing $aur_pkg..."
      makepkg $makepkg_type
      cd ..
      rm -rf $aur_pkg
    done
  elif [[ $1 == "-Spf" ]]; then
    sudo pacman -S $pkg_string
  fi
# SAH Install Local/Remote Package
elif [[ $1 == "-U" ]]; then
  sudo pacman -U $2
# SAH Update Package Database
elif [[ $1 == "-Syy" || $1 == "update" ]]; then
  sudo pacman -Syy
# SAH Update
elif [[ $1 == "-Syu" || $1 == "upgrade" ]]; then
  if [[ $update_pacman_check == "true" ]]; then
    echo "Checking for updates from Pacman..."
    sudo pacman -Syu
  elif [[ $update_pacman_check == "false" ]]; then
    echo "Updating of Pacman packages is disabled."
  fi

  if [[ $update_aur_check == "true" ]]; then
    echo
    echo "Checking for updates from AUR..."

    sudo pacman -Qqm > $pkg_list_path
    sudo pacman -Qm > $pkg_list_path_v

    readarray -t pkg_list < $pkg_list_path
    readarray -t pkg_list_v < $pkg_list_path_v
    pkgz=${#pkg_list[@]}
    pkgz_v=${#pkg_list_v[@]}

    aur_update_ignore_count=$(cat $SAH_config_path | grep "aur_update_ignore" | awk -F "=" '{print $2}' | tr ',' '\n' | wc -l)
    for (( i = 0; i < $aur_update_ignore_count; i++ )); do
      aur_update_ignore_num=$(($i + 1))
      aur_update_ignore[$i]=$(cat $SAH_config_path | grep "aur_update_ignore" | awk -F "=" '{print $2}' | awk -F "," "{print \$$aur_update_ignore_num}")
    done

    mkdir $PKGBUILDs_path
    for (( i = 0; i < $pkgz; i++ )); do
      check_pkg=${pkg_list[$i]}
      check_pkg_v=${pkg_list_v[$i]}
      check_pkg_v=$(echo $check_pkg_v | awk '{print $2}')
      check_pkg_num=$(($i + 1))

      latest_version_message="-> [$check_pkg_num / $pkgz] $check_pkg - you have the latest version."
      update_message="-> [$check_pkg_num / $pkgz] Updating $check_pkg..."
      aur_update_ignore_message="-> [$check_pkg_num / $pkgz] $check_pkg - skipped."
      aur_update_notify_message="-> [$check_pkg_num / $pkgz] $check_pkg - new version is available."

      # Exceptions
      if [[ $check_pkg != "sah" ]]; then
        wget_link="https://aur.archlinux.org/cgit/aur.git/plain/PKGBUILD?h=$check_pkg"
        git_clone_link="https://aur.archlinux.org/$check_pkg.git"
      elif [[ $check_pkg == "sah" ]]; then
        wget_link="https://raw.githubusercontent.com/zurg3/$check_pkg/master/PKGBUILD"
        git_clone_link="https://github.com/zurg3/$check_pkg.git"
      fi

      wget -q $wget_link -O $PKGBUILDs_path/$check_pkg.txt

      version_main=$(cat $PKGBUILDs_path/$check_pkg.txt | grep "pkgver" | head -n 1 | awk -F "=" '{print $2}')
      version_patch=$(cat $PKGBUILDs_path/$check_pkg.txt | grep "pkgrel" | head -n 1 | awk -F "=" '{print $2}')
      version_full="$version_main-$version_patch"

      if [[ " ${aur_update_ignore[*]} " != *" $check_pkg "* ]]; then
        if [[ $check_pkg_v == $version_full ]]; then
          echo "$latest_version_message"
          if [[ $check_pkg != "sah" && $auto_cache_check == "true" ]]; then
            test -d $SAH_cache_path/$check_pkg
            if [[ $? == "0" ]]; then
              cd $SAH_cache_path/$check_pkg
              git pull
              cd - > /dev/null
            elif [[ $? == "1" ]]; then
              git clone -q $git_clone_link $SAH_cache_path/$check_pkg
            fi
          fi
        elif [[ $check_pkg_v != $version_full ]]; then
          # Version from PKGBUILD may has the single quotes.
          echo "$version_full" | grep -q "'"
          if [[ $? == "0" ]]; then
            pkgver_sq=$(echo "$check_pkg_v" | awk -F "-" '{print $1}')
            pkgrel_sq=$(echo "$check_pkg_v" | awk -F "-" '{print $2}')
            check_pkg_v_sq="'$pkgver_sq'-'$pkgrel_sq'"
            if [[ $check_pkg_v_sq != $version_full ]]; then
              if [[ $aur_update_notify_check == "true" ]]; then
                echo "$aur_update_notify_message"
              elif [[ $aur_update_notify_check == "false" ]]; then
                echo "$update_message"
                git clone $git_clone_link
                if [[ $check_pkg != "sah" && $auto_cache_check == "true" ]]; then
                  test -d $SAH_cache_path/$check_pkg
                  if [[ $? == "0" ]]; then
                    cd $SAH_cache_path/$check_pkg
                    git pull
                    cd - > /dev/null
                  elif [[ $? == "1" ]]; then
                    git clone -q $git_clone_link $SAH_cache_path/$check_pkg
                  fi
                fi
                cd $check_pkg
                makepkg $makepkg_type
                cd ..
                rm -rf $check_pkg
              fi
            elif [[ $check_pkg_v_sq == $version_full ]]; then
              echo "$latest_version_message"
            fi
          elif [[ $? == "1" ]]; then
            if [[ $aur_update_notify_check == "true" ]]; then
              echo "$aur_update_notify_message"
            elif [[ $aur_update_notify_check == "false" ]]; then
              echo "$update_message"
              git clone $git_clone_link
              if [[ $check_pkg != "sah" && $auto_cache_check == "true" ]]; then
                test -d $SAH_cache_path/$check_pkg
                if [[ $? == "0" ]]; then
                  cd $SAH_cache_path/$check_pkg
                  git pull
                  cd - > /dev/null
                elif [[ $? == "1" ]]; then
                  git clone -q $git_clone_link $SAH_cache_path/$check_pkg
                fi
              fi
              cd $check_pkg
              makepkg $makepkg_type
              cd ..
              rm -rf $check_pkg
            fi
          fi
        fi
      elif [[ " ${aur_update_ignore[*]} " == *" $check_pkg "* ]]; then
        echo "$aur_update_ignore_message"
      fi
    done
    rm $pkg_list_path
    rm $pkg_list_path_v
    rm -rf $PKGBUILDs_path
  elif [[ $update_aur_check == "false" ]]; then
    echo
    echo "Updating of AUR packages is disabled."
  fi
# SAH Clean
elif [[ $1 == "-Sc" || $1 == "autoremove" ]]; then
  sudo pacman -Sc
# SAH Remove
elif [[ $1 == "-R" || $1 == "remove" ]]; then
  sudo pacman -R ${@:2}
# SAH Remove With Dependencies
elif [[ $1 == "-Rs" || $1 == "purge" ]]; then
  sudo pacman -Rs ${@:2}
# SAH Remove Orphans
elif [[ $1 == "-Ro" ]]; then
  sudo pacman -R $(sudo pacman -Qdtq)
# SAH Installed All
elif [[ $1 == "-Q" || $1 == "list" ]]; then
  query_pkg_count=$(sudo pacman -Q | wc -l)
  echo "Installed packages (All) [$query_pkg_count packages]:"
  sudo pacman -Q
# SAH Installed Explicitly
elif [[ $1 == "-Qe" ]]; then
  query_pkg_count=$(sudo pacman -Qe | wc -l)
  echo "Installed packages (Explicitly) [$query_pkg_count packages]:"
  sudo pacman -Qe
# SAH Installed AUR
elif [[ $1 == "-Qm" ]]; then
  query_pkg_count=$(sudo pacman -Qm | wc -l)
  echo "Installed packages (AUR) [$query_pkg_count packages]:"
  sudo pacman -Qm
# SAH Search
elif [[ $1 == "-Ss" ]]; then
  sudo pacman -Ss $2
# SAH Search Installed
elif [[ $1 == "-Qs" ]]; then
  sudo pacman -Qs $2
# SAH Show Info
elif [[ $1 == "-Si" ]]; then
  sudo pacman -Si $2
# SAH Show Info Installed
elif [[ $1 == "-Qi" ]]; then
  sudo pacman -Qi $2
# SAH Orphans
elif [[ $1 == "-Qdt" ]]; then
  query_pkg_count=$(sudo pacman -Qdt | wc -l)
  echo "Packages no longer required as dependencies (orphans) [$query_pkg_count packages]:"
  sudo pacman -Qdt
# SAH Custom Pacman Operation
elif [[ $1 == "custom" ]]; then
  sudo pacman ${@:2}
# SAH Statistics
elif [[ $1 == "stat" ]]; then
  query_pkg_count_Q=$(sudo pacman -Q | wc -l)
  query_pkg_count_Qe=$(sudo pacman -Qe | wc -l)
  query_pkg_count_Qm=$(sudo pacman -Qm | wc -l)
  query_pkg_count_Qdt=$(sudo pacman -Qdt | wc -l)

  echo "Packages statistics"

  echo "Installed packages (All): $query_pkg_count_Q"
  echo "Installed packages (Explicitly): $query_pkg_count_Qe"
  echo "Installed packages (AUR): $query_pkg_count_Qm"
  echo "Packages no longer required as dependencies (orphans): $query_pkg_count_Qdt"
# SAH Config
elif [[ $1 == "config" ]]; then
  sudo $SAH_editor $SAH_config_path
# SAH Reset Config
elif [[ $1 == "resetconfig" ]]; then
  sudo curl -s "https://raw.githubusercontent.com/zurg3/sah/v$VERSION/sah_config_default" -o $SAH_config_path
  sudo $SAH_editor $SAH_config_path
# SAH Pacman Config
elif [[ $1 == "pacconf" ]]; then
  sudo $SAH_editor $pacman_config_path
# SAH Pacman Mirrorlist
elif [[ $1 == "mirrorlist" ]]; then
  sudo $SAH_editor $mirrorlist_path
# SAH Pacman Update Mirrors
elif [[ $1 == "updatemirrors" ]]; then
  sudo curl -s "https://archlinux.org/mirrorlist/?country=$mirrorlist_country&protocol=$mirrorlist_protocol&ip_version=$mirrorlist_ip_version" -o $mirrorlist_path
  sudo $SAH_editor $mirrorlist_path
# SAH Changelog
elif [[ $1 == "changelog" ]]; then
  $SAH_viewer $SAH_changelog_path
# SAH AUR TOP Packages
elif [[ $1 == "top" ]]; then
  wget -q https://aur.archlinux.org/packages -O - | sed -n "/<table class=\"results\">/,/<\/table>/p" > /tmp/aur_packages.html
  date -u +"%d.%m.%Y" > $aur_top_path
  date -u +"%H:%M" >> $aur_top_path
  for (( i = 1; i <= 10; i++ )); do
    xmllint --xpath "//table//tbody//tr[$i]//td[1]//a/text()" /tmp/aur_packages.html >> $aur_top_path
  done

  readarray -t aur_top < $aur_top_path
  aur_top_date=${aur_top[0]}
  aur_top_time=${aur_top[1]}
  if [[ $2 == "" ]]; then
    aur_top_num=1
    echo "TOP-10 popular AUR packages (updated on $aur_top_date at $aur_top_time UTC):"
    for (( i = 2; i < 12; i++ )); do
      echo "$aur_top_num. ${aur_top[$i]}"
      aur_top_num=$(($aur_top_num + 1))
    done
  elif [[ $2 != "" ]]; then
    if (( $2 >= 1 && $2 <= 10 )); then
      aur_top_install=$(($2 + 1))
      aur_pkg=${aur_top[$aur_top_install]}
      git clone https://aur.archlinux.org/$aur_pkg.git
      if [[ $auto_cache_check == "true" ]]; then
        test -d $SAH_cache_path/$aur_pkg
        if [[ $? == "0" ]]; then
          cd $SAH_cache_path/$aur_pkg
          git pull
          cd - > /dev/null
        elif [[ $? == "1" ]]; then
          git clone -q https://aur.archlinux.org/$aur_pkg.git $SAH_cache_path/$aur_pkg
        fi
      fi
      cd $aur_pkg
      echo "Installing $aur_pkg..."
      makepkg $makepkg_type
      cd ..
      rm -rf $aur_pkg
    else
      echo "You can install packages only in the range from 1 to 10!"
    fi
  fi
# SAH Cache
elif [[ $1 == "cache" ]]; then
  if [[ $2 != "install" && $2 != "update" && $2 != "remove" && $2 != "list" && $2 != "clean" ]]; then
    aur_pkg_range="${@:2}"
    for aur_pkg in $aur_pkg_range
    do
      test -d $SAH_cache_path/$aur_pkg
      if [[ $? == "1" ]]; then
        echo "Caching $aur_pkg..."
        git clone https://aur.archlinux.org/$aur_pkg.git $SAH_cache_path/$aur_pkg
      fi
    done
  fi

  aur_pkg_range="${@:3}"
  if [[ $2 == "install" ]]; then
    for aur_pkg in $aur_pkg_range
    do
      test -d $SAH_cache_path/$aur_pkg
      if [[ $? == "0" ]]; then
        cd $SAH_cache_path/$aur_pkg
        echo "Installing $aur_pkg..."
        makepkg $makepkg_type
      elif [[ $? == "1" ]]; then
        echo "$aur_pkg - package not found."
      fi
    done
  elif [[ $2 == "update" ]]; then
    if [[ $3 != "--all" ]]; then
      for aur_pkg in $aur_pkg_range
      do
        test -d $SAH_cache_path/$aur_pkg
        if [[ $? == "0" ]]; then
          cd $SAH_cache_path/$aur_pkg
          echo "Updating $aur_pkg..."
          git pull
        elif [[ $? == "1" ]]; then
          echo "$aur_pkg - package not found."
        fi
      done
    elif [[ $3 == "--all" ]]; then
      ls -1 $SAH_cache_path > $SAH_cache_list_path
      readarray -t cache_list < $SAH_cache_list_path
      cache_list_num=${#cache_list[@]}
      for (( i = 0; i < $cache_list_num; i++ )); do
        test -d $SAH_cache_path/${cache_list[$i]}
        if [[ $? == "0" ]]; then
          cd $SAH_cache_path/${cache_list[$i]}
          echo "Updating ${cache_list[$i]}..."
          git pull
        elif [[ $? == "1" ]]; then
          echo "${cache_list[$i]} - package not found."
        fi
      done
    fi
  elif [[ $2 == "remove" ]]; then
    for aur_pkg in $aur_pkg_range
    do
      test -d $SAH_cache_path/$aur_pkg
      if [[ $? == "0" ]]; then
        rm -rf $SAH_cache_path/$aur_pkg
        echo "Removed $aur_pkg..."
      elif [[ $? == "1" ]]; then
        echo "$aur_pkg - package not found."
      fi
    done
  elif [[ $2 == "list" ]]; then
    ls -1 $SAH_cache_path
  elif [[ $2 == "clean" ]]; then
    rm -rf $SAH_cache_path/*
  fi
# SAH Browse
elif [[ $1 == "browse" ]]; then
  if [[ $SAH_browser == "" ]]; then
    echo "You need specify a web browser in SAH config!"
  elif [[ $SAH_browser != "" ]]; then
    if [[ $2 == "" ]]; then
      echo "Usage: sah browse [--pacman|--aur] or sah browse [--pacman|--aur] [package]"
    elif [[ $2 != "" ]]; then
      if [[ $3 == "" ]]; then
        if [[ $2 == "--pacman" ]]; then
          $SAH_browser https://archlinux.org/packages/
        elif [[ $2 == "--aur" ]]; then
          $SAH_browser https://aur.archlinux.org/packages/
        else
          echo "Usage: sah browse [--pacman|--aur] or sah browse [--pacman|--aur] [package]"
        fi
      elif [[ $3 != "" ]]; then
        if [[ $2 == "--pacman" ]]; then
          $SAH_browser https://archlinux.org/packages/?q=$3
        elif [[ $2 == "--aur" ]]; then
          $SAH_browser https://aur.archlinux.org/packages/$3
        else
          echo "Usage: sah browse [--pacman|--aur] or sah browse [--pacman|--aur] [package]"
        fi
      fi
    fi
  fi
# SAH View PKGBUILD
elif [[ $1 == "view" ]]; then
  mkdir $PKGBUILDs_path
  wget_link="https://aur.archlinux.org/cgit/aur.git/plain/PKGBUILD?h=$2"
  wget -q $wget_link -O $PKGBUILDs_path/$2.txt
  $SAH_viewer $PKGBUILDs_path/$2.txt
  rm -rf $PKGBUILDs_path
# SAH Version
elif [[ $1 == "--version" || $1 == "-V" ]]; then
  echo "Simple AUR Helper (SAH) and Pacman wrapper v$VERSION"
# SAH Help
elif [[ $1 == "" || $1 == "--help" || $1 == "-h" ]]; then
  $SAH_viewer $SAH_help_path
# SAH Debug
elif [[ $1 == "debug" ]]; then
  # Man page preview: nroff -man sah.8 | less

  version_length=${#VERSION}
  if [[ $version_length == "3" ]]; then
    debug_art="=========================="
  elif [[ $version_length == "5" ]]; then
    debug_art="============================"
  elif [[ $version_length == "6" ]]; then
    debug_art="============================="
  elif [[ $version_length == "7" ]]; then
    debug_art="=============================="
  else
    debug_art="= = = = ="
  fi

  echo "===== SAH Debug v$VERSION ====="
  echo "$debug_art"
  echo "Arch Linux $kernel_version"
  echo "$USER@$HOSTNAME"
  echo "$date_time_format"
  echo "$debug_art"
else
  echo "Something is wrong!"
fi
